### PR TITLE
LDrawLoader: Group object by Parts and smooth normals

### PR DESCRIPTION
**Pull request by gkjohnson**
This PR groups geometry by [LDraw "Part"](https://www.ldraw.org/article/218.html) type which provides a more intuitive grouping of triangles (a "brick" is single piece of geometry representing a brick vs a collection of primitive geometric pieces), which could lead to other caching and geometry reuse in the future and adds support for smoothing normals and retaining hard edges where line segments are defined.

image

It should be noted that smoothing normals adds significant overhead to the parsing and can be disabled by setting `loader.smoothNormals = false`. Maybe it can be improved in the future.

This PR reorganizes a few parts of the code to accomodate normal smoothing and adds support for displaying conditional lines, as well.

For testing:
https://raw.githack.com/gkjohnson/three.js/ldraw-group-parts/examples/webgl_loader_ldraw.html

Includes changes from #16383.

@yomboprime

### CODE REVIEW

**Comment by yomboprime** [comment]
The smoothed normals are looking very good!

I've only seen the demo, later I will be reviewing your code, but for now:

- I can't see the envMap if I enable it.
- Conditional lines are supposed to be drawn only if a complex condition about the line and two additional points is met, in screen space. This is non trivial and requires custom shaders and probably custom attributes. But I guess it is ok to draw them without the condition right now (which is what you've done). This algorithm with the complex condition was developed in the nineties, and the approach right now would be rather like disabling all the lines and applying a toon shader.

**Comment by gkjohnson** [comment]
> I can't see the envMap if I enable it.

I think this is because of the changing to use `MeshStandardMaterial` in the previous PR. The "roughness" is set to 0.3 on brick materials now which dilutes the environment map a bit.

> Conditional lines are supposed to be drawn only if a complex condition about the line and two additional points is met, in screen space. ...

Yeah I have another PR in progress that adds conditional lines as expected using a custom shader. I originally intended to use them here for smoothing but wound up taking a different route. They're hidden by default on load, though. I can remove them in this PR if that's preferred!

**Comment by yomboprime** [comment]
All right, I approve these changes (but it depends on #16383 being merged)

File: examples/js/loaders/LDrawLoader.js [hunk 1]
```diff
@@ -1270,11 +1541,14 @@ THREE.LDrawLoader = ( function () {
...
 					// Line type 2: Line segment
+					// Line type 5: Conditional Line segment
 					case '2':
+					case '5':
 
 						var material = parseColourCode( lp, true );
+						var arr = lineType === '2' ? lineSegments : conditionalSegments;
 
-						lineSegments.push( {
+						arr.push( {
```
Comment (mrdoob) [inline examples/js/loaders/LDrawLoader.js]: Pirate var 😁

**Comment by mrdoob** [comment]
Thanks!

**Pull request merged**